unittest/client/src/ilm_control_unittests.cpp

```diff
@@ -0,0 +1,5071 @@
+ * @brief               Test case of ilm_layerSetRenderOrder() where ilm context is initilized and valid layer id
+ * @test_procedure Steps:
+ *                      -# Calling the ilm_layerSetRenderOrder()
+ *                      -# Verification point:
+ *                         +# ilm_layerSetRenderOrder() must return ILM_SUCCESS
+ *                         +# wl_proxy_marshal_flags() must be called {1+l_number} time
+ *                         +# wl_display_flush() must be called once time
+ */
+TEST_F(IlmControlTest, ilm_layerSetRenderOrder_success)
+{
+    t_ilm_layer l_layerId = 100;
+    t_ilm_int l_number = MAX_NUMBER;
+    ASSERT_EQ(ILM_SUCCESS, ilm_layerSetRenderOrder(l_layerId, mp_ilmSurfaceIds, l_number));
+
+    ASSERT_EQ(1 + l_number, wl_proxy_marshal_flags_fake.call_count);
+    ASSERT_EQ(1, wl_display_flush_fake.call_count);
+}
+
+/** ================================================================================================
+ * @test_id             ilm_surfaceSetVisibility_wrongCtx
+ * @brief               Test case of ilm_surfaceSetVisibility() where ilm context wl.controller is null object
+ * @test_procedure Steps:
+ *                      -# Set the ilm context wl.controller is null object
+ *                      -# Calling the ilm_surfaceSetVisibility()
+ *                      -# Verification point:
+ *                         +# ilm_surfaceSetVisibility() must return ILM_FAILED
+ *                         +# wl_proxy_marshal_flags() not be called
+ *                         +# wl_display_flush() not be called
+ */
+TEST_F(IlmControlTest, ilm_surfaceSetVisibility_wrongCtx)
+{
+    ilm_context.wl.controller = nullptr;
+
+    t_ilm_surface l_surfaceId = 1;
+    t_ilm_bool l_newVisibility = ILM_FALSE;
+    ASSERT_EQ(ILM_FAILED, ilm_surfaceSetVisibility(l_surfaceId, l_newVisibility));
+
+    ASSERT_EQ(0, wl_proxy_marshal_flags_fake.call_count);
+    ASSERT_EQ(0, wl_display_flush_fake.call_count);
+}
+
+/** ================================================================================================
+ * @test_id             ilm_surfaceSetVisibility_success
+ * @brief               Test case of ilm_surfaceSetVisibility() where ilm context is initilized and valid surface id
+ * @test_procedure Steps:
+ *                      -# Calling the ilm_surfaceSetVisibility()
+ *                      -# Verification point:
+ *                         +# ilm_surfaceSetVisibility() must return ILM_SUCCESS
+ *                         +# wl_proxy_marshal_flags() must be called once time
+ *                         +# wl_display_flush() must be called once time
+ */
+TEST_F(IlmControlTest, ilm_surfaceSetVisibility_success)
+{
+    t_ilm_surface l_surfaceId = 1;
+    t_ilm_bool l_newVisibility = ILM_TRUE;
+    ASSERT_EQ(ILM_SUCCESS, ilm_surfaceSetVisibility(l_surfaceId, l_newVisibility));
+
+    ASSERT_EQ(1, wl_proxy_marshal_flags_fake.call_count);
+    ASSERT_EQ(1, wl_display_flush_fake.call_count);
```

Comment:
Should we check for arguments passed to wl_proxy_marshal_flags_fake (like if we this call was made with the right opcode[IVI_WM_SET_SURFACE_VISIBILITY] and correct surface id & visibility)

    ASSERT_EQ(IVI_WM_SET_SURFACE_VISIBILITY, wl_proxy_marshal_flags_fake.arg1_history[0]);
    ASSERT_EQ(l_surfaceId, wl_proxy_marshal_flags_fake.arg5_history[0]);
    ASSERT_EQ(newVisibility, wl_proxy_marshal_flags_fake.arg6_history[0]);

Reply:
Also if version information can be checked?

Reply:
Above should be taken care in all test case where we have asserted success and we expect wl_proxy_marshal_flags_fake() to be called.

Reply:
Applied the checking for the second argument for all tests are calling `wl_proxy_marshal_flags_fake`. `wl_proxy_marshal_flags` is a Variadic function, so we only check for first five arguments (0->4). so it means, we can't check the data of request.

Reply:
ohh, yes!!
But I found something here: https://github.com/meekrosoft/fff/issues/57
Might be we can use custom fake function?

Reply:
Or atleast 
`ASSERT_EQ(IVI_WM_SET_SURFACE_VISIBILITY, wl_proxy_marshal_flags_fake.arg1_history[0]);`
can be verified.

Reply:
> Or atleast `ASSERT_EQ(IVI_WM_SET_SURFACE_VISIBILITY, wl_proxy_marshal_flags_fake.arg1_history[0]);` can be verified.

This thing is done.

Reply:
> ohh, yes!! But I found something here: [meekrosoft/fff#57](https://github.com/meekrosoft/fff/issues/57) Might be we can use custom fake function?

I need to look more. thanks for the link.